LibGfx/ILBM: Add an IFF-ILBM decoder :)

warpdesign opened the pull request:
IFF was a generic container fileformat that was popular on the Amiga & DOS since it was the only file format supported by Deluxe Paint. [ILBM](https://wiki.amigaos.net/wiki/ILBM_IFF_Interleaved_Bitmap) is an image format popular in the late eighties/nineties that uses the IFF container.

This is a very first version of the decoder that only supports _ByteRun_ compressed files with bpp <= 8.

Only the minimal chunks are decoded: CMAP, BODY, BMHD.

I am planning to add support for the following variants:

- EHB (32 colours + lighter 32 colours)
- HAM6 / HAM8 (special mode that allowed to display the whole Amiga 4096 colours / 262 144 colours palette)
- TrueColor (24bit)
- Mask/Transparent Color support

Things that could be fun to do:

- Still images could be animated using color cycle information

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+static ErrorOr<Vector<u32>> decode_cmap_chunk(Chunk cmap_chunk)
+{
+    auto data = cmap_chunk.data;
+    const u8 size = (u8)(data.size() / 3);
```
We only use west `const` in this codebase :)
And avoid C-cast

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+{
+    if (index > 0)
+        return Error::from_string_literal("ILBMImageDecoderPlugin: frame index must be 0");
+    else {
```
No need for `else` after a `return`

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+        if (m_context->state < ILBMLoadingContext::ChunksDecoded)
+            TRY(decode_iff_chunks(*m_context));
+
+        RefPtr<Gfx::Bitmap> bitmap = TRY(chunky_to_bitmap(*m_context));
```
That will be done every time the frame is requested, you only need to do it once

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+    // Current file color_table in BGRA8888 format
+    Vector<u32> color_table;
+
+    ByteBuffer pixel_data;
```
You're basically storing the image twice, once in `pixel_data` and once in the bitmap. I'm pretty sure it can be avoided

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+    auto width = context.bm_header.width;
+    auto height = context.bm_header.height;
+    auto color_table = context.color_table;
+    auto chunky = context.pixel_data;
```
In a general manner, I'm not a fan of these aliases. I think that they hide information (in fact, where does the information come from) just to save a few characters. But that's more a question of style, so do what you want.
However, you are copying the whole bitmap here, and that's something I can argue on being bad :)

warpdesign replied:
It's not only to save characters but also to save some indirections: I guess that's a bad habit from the JS world. Not sure that's needed in Cpp?

LucasChollet replied:
1. Then you're doing premature optimization, don't try to optimize stuff without needing it and measuring it. Having a readable code is necessary, optimizing it comes after.
2. C++ compilers are very good at optimizing stuff. By doing stuff like this, you have more chance to mess with their optimizations than to outsmart them.
3. Using `auto&` would create a local reference (the size of a pointer) and that's the closest to what you want. Using `auto` and `=` calls the copy constructor which makes a __deep__ copy, copying a whole image, pixel by pixel, to save an indirection probably won't make you save some time.

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+#include <AK/FixedArray.h>
+#include <AK/Format.h>
+#include <AK/MemoryStream.h>
+#include <AK/String.h>
```
Is this used?

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h:
```diff
@@ -0,0 +1,33 @@
+
+#pragma once
+
+#include <AK/MemoryStream.h>
```
That's included both here and in the .cpp but not used

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h:
```diff
@@ -0,0 +1,33 @@
+
+private:
+    ILBMImageDecoderPlugin(ReadonlyBytes, OwnPtr<ILBMLoadingContext>);
+    OwnPtr<ILBMLoadingContext> m_context;
```
Use a NonnullOwnPtr

LucasChollet reviewed (commented):
Thanks for adding a decoder!

There are a lot of things, but most are trivial.

It would be really nice to add both a test and a fuzzer.

warpdesign commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,373 @@
+
+namespace Gfx {
+
+struct FourCC {
```
Indeed: I borrowed this struct. If I want to reuse it, how am I supposed to share it? What's the habits in cpp/serenityOS to share pieces of code like this?

LucasChollet replied:
Put it in a `.h`, you can probably drop it at the end of `ImageDecoder.h` so every decoder can use it.
We will move it to a utility file if we have more things to share later on.

warpdesign replied:
I put FourCC inside `ImageDecoder.h` and also modified `WebPLoader.cpp`: where FourCC was originally defined, not sure this is the way to do.

warpdesign commented:
> Thanks for adding a decoder!
> 
> There are a lot of things, but most are trivial.
> 
> It would be really nice to add both a test and a fuzzer.

I fixed most problems: only two issues remain, see my comments.

As for the the test & fuzzer: could you point me some example as a reference?

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
-        u32 color = (data[i * 3] << 16) | (data[(i * 3) + 1] << 8) | data[(i * 3) + 2];
-        color |= 0xff000000;
-        color_table.append(color);
+        color_table[i] = Color(data[i * 3], data[(i * 3) + 1], data[(i * 3) + 2], 255);
```
`Color` has a construct that only takes three parameters

warpdesign replied:
I saw that, but I was already thinking about using it when I'll have implemented mask/transparent color. But I guess I can switch to this constructor when I'll be implementing that.

LucasChollet replied:
You will need to modify the line anyway, no need to bloat it right now

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -97,17 +90,11 @@ struct ILBMLoadingContext {
-    BitmapHeader bm_header;
-
-    Optional<IntSize> size;
+    BitmapHeader BMHDHeader;
```
I was more thinking of `BMHDHeader bm_header;`
Variable names shouldn't be capitalized

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
     auto data = cmap_chunk.data;
-    const u8 size = (u8)(data.size() / 3);
-    Vector<u32> color_table;
+    u8 const size = (data.size() / 3);
```
Can you also remove both parentheses?

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
 }
 
-static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context)
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context, ByteBuffer& chunky)
```
use a `const&` if you are only reading from it

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -154,20 +137,22 @@ static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context
         for (int col = 0; col < width; ++col) {
             u8 index = chunky[(width * row) + col];
-            bitmap->scanline(row)[col] = color_table[index];
+            bitmap->scanline(row)[col] = context.color_table[index].value();
```
`Bitmap` has a `set_pixel` method that can take a `Color` directly

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -154,20 +137,22 @@ static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context
-    u16 w = context.bm_header.width;
-    u16 h = context.bm_header.height;
-    u8 planes = context.bm_header.planes;
+    u16 w = context.BMHDHeader.width;
```
> Use full words, except in the rare case where an abbreviation would be more canonical and easier to understand.

[CodingStyle.md](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#right-1)

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -225,12 +210,15 @@ static ErrorOr<void> decode_body_chunk(Chunk body_chunk, ILBMLoadingContext& con
+        auto pixel_data = TRY(planar_to_chunky(plane_data, context));
+
+        auto bitmap = TRY(chunky_to_bitmap(context, pixel_data));
+        context.bitmap = bitmap;
```
No need for these two steps
```cpp
context.bitmap = TRY(chunky_to_bitmap(context, pixel_data));
```

warpdesign replied:
If you think I can directly write to the bitmap, I don't think so: I really need to de-interleave the bitplanes first, then only I can write to the bitmap. So I can put both steps into the same method, but these steps can't be put into a single step.

LucasChollet replied:
I was speaking of the two-step assignment, the temporary `bitmap` variable is useless.

```cpp
auto bitmap = TRY(chunky_to_bitmap(context, pixel_data));
context.bitmap = bitmap;
```
is just
```cpp
context.bitmap = TRY(chunky_to_bitmap(context, pixel_data));
```
but with extra steps

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -278,14 +266,13 @@ static ErrorOr<void> decode_iff_chunks(ILBMLoadingContext& context)
             TRY(decode_body_chunk(chunk, context));
-        else if (chunk.type == FourCC("CRNG"))
+            context.state = ILBMLoadingContext::State::BitmapDecoded;
+        } else if (chunk.type == FourCC("CRNG"))
```
https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#omission-of-curly-braces-from-statement-blocks

You're not respecting this in `decode_body_chunk`

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
```
I still maintain that having both `HeaderDecoded` and `BitmapHeaderDecoded` is useless
No user should have a ILBMDecoder with its context state set to something before both the file header and the bitmap header are decoded.

LucasChollet replied:
Third time's the charm, maybe?

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,373 @@
+    ReadonlyBytes data;
+};
+
+enum CompressionType : u8 {
```
> (that's valid everywhere)

Please do it for *every* `enum`

`enum class` do not pollute the global namespace, so you can't confuse a common name that is shared between two of them. You're manually adding a prefix to each `enum` as a workaround this issue. So please use `enum class` and drop the prefix

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+
+    for (u16 y = 0; y < h; y++) {
+        for (u8 p = 0; p < planes; p++) {
+            const u8 plane_mask = 1 << p;
```
That's also an east const

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,372 @@
+
+bool ILBMImageDecoderPlugin::sniff(ReadonlyBytes data)
+{
+
```
Can you remove this line?

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -300,23 +287,23 @@ static ErrorOr<void> decode_ilbm_first_chunk(ILBMLoadingContext& context)
+        context.BMHDHeader.x,
+        context.BMHDHeader.y,
+        context.BMHDHeader.planes,
+        (u8)context.BMHDHeader.mask,
```
Same as before, avoid C-cast
For enums we have `to_underlying`

LucasChollet reviewed (commented):
2nd round!

When you modify your code after someone requested changes, fixup them in the corresponding commit instead of pushing a new one. (ask on discord if you face issues doing it)

warpdesign commented:
@LucasChollet I think that's it (at least for this second round :)) Do you want me to squash both commits?

LucasChollet commented:
I only had a quick glance but it looks like you did not remove the prefixes from your `enum`s.

Yes, please squash them (as the English term), but use git's fixup feature.

Put the FourCC extraction in its own commit (I indeed intended you to remove it from the WebP decoder). No need to include `ImageDecoder.h` from every `*Loader.cpp` file, it's already done but including the corresponding `*Loader.h`.

warpdesign commented:
> I only had a quick glance but it looks like you did not remove the prefixes from your `enum`s.
> 
> Yes, please squash them (as the English term), but use git's fixup feature.
> 
> Put the FourCC extraction in its own commit (I indeed intended you to remove it from the WebP decoder). No need to include `ImageDecoder.h` from every `*Loader.cpp` file, it's already done but including the corresponding `*Loader.h`.

Ok. I'll do it in a separate commit.

What do you mean by "remove the prefixes"?

LucasChollet commented:
As an example all values in CompressionType start with "Cmp_" that's useful with a plane `enum` but useless with an `enum class` so please remove them

warpdesign commented:
I moved the `FourCC` changes into a separate commit then squashed all ILBM-related ones into one commit.

I also removed the prefixes in `enum class`.

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,338 @@
+
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context, ByteBuffer const& chunky)
+{
+    auto& width = context.bm_header.width;
```
The basic rule of thumb for should I take a reference or a copy to an object is: if it fits into a register a copy is cheaper, take a reference otherwise.
Integers fall into the first category, don't take a reference if it's not needed.

Also, when I see someone taking a mutable reference, I expect it to be used to be written to. Explicitly taking a `const` reference makes your intentions much clearer.

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,338 @@
+    dbgln_if(ILBM_DEBUG, "decode_body_chunk {}", body_chunk.data.size());
+
+    if (context.bm_header.compression == CompressionType::ByteRun) {
+        auto& data = body_chunk.data;
```
Hmm, what is this line doing?

warpdesign replied:
Removed it.

LucasChollet replied:
I can still see it :eyes:

warpdesign replied:
 Oops, should be really gone now.

LucasChollet commented on Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:
```diff
@@ -0,0 +1,338 @@
+    return {};
+}
+
+static ErrorOr<void> decode_ilbm_first_chunk(ILBMLoadingContext& context)
```
If the first chunk is named BMHD, why not naming the function `decode_bmhd_chunk` (or something else if you want, just more explicit than the current name)

LucasChollet commented on Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:
```diff
@@ -0,0 +1,18 @@
+/*
```
From [CONTRIBUTING](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy):
> Write the commit message subject line in the imperative mood ("Foo: Change the way dates work", not "Foo: Changed the way dates work").

And this hasn't much to do with `LibGfx`, you can make two separate commits or otherwise use `Meta+Tests` for the category.

warpdesign replied:
I used `Meta+Tests` for the category.

LucasChollet replied:
You still don't use the imperative mood for this commit.
"Meta+Tests: Add a fuzzer and a test for the ILBM decoder" would be better

warpdesign replied:
Done :)

LucasChollet commented on Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
```
I think that the year can be bumped

LucasChollet reviewed (commented):
Sorry for the delay, last nitpicks, and then it should be good to go!

When you rebase your changes, make sure to keep the list in `Base/res/apps/ImageViewer.af` sorted.

warpdesign commented on Meta/gn/secondary/AK/BUILD.gn:
```diff
@@ -283,6 +283,7 @@ write_cmake_config("ak_debug_gen") {
     "HTTPJOB_DEBUG=",
     "HUNKS_DEBUG=",
     "ICO_DEBUG=",
+    "ILBM_DEBUG=",
```
I have added it in `Meta/gn/secondary/Userland/Libraries/LibGfx/BUILD.gn`: not sure if it's the only place where it's referenced.
